text/0113-follower-read-cache.md

```diff
@@ -0,0 +1,79 @@
+
+## Motivation
+
+In a cloud environment, customers often prefer to perform follower reads within the same AZ to avoid incurring cross-AZ network costs. However, each follower read requires sending a read index request to the leader to check for any pending write locks in memory. These RPC calls can be quite expensive and can significantly increase CPU utilization, especially in read-heavy traffic scenarios. This increase in CPU cost eliminates the cost savings achieved by avoiding cross-AZ traffic.
+To mitigate this issue in read-heavy environments, it is not necessary to send a read index request for every single follower read request. This is because, in 99% of cases, there are no pending locks. 
+
+## Detailed design
+
+safe-ts in follower is used to determine the minimum timestamp that guarantees the consistent reads from replica. Currently, safe-ts is advanced by advance-ts message from the leader. In this proposal, response of read index messsages can also be used to advance safe-ts in follower if there are no pending pre writes on leader. 
+Changes to advance safe-ts from raft read index response is pretty straightforward and most of the existing logic in the stale read path will be reused. 
+Ref counters per region need to be added in the leader to keep track of total number of pending locks and lock status need to be included in a raft read index response. These are the detailed steps on how a follower read works
+- Follower reads will initially compare the safe-ts with the timestamp in the request, and only if the request timestamp is greater than the safe-ts, a read index request will be sent.
+- Read index request on leader check the lock reference counter and send its status through raft message in read index response
+- Follower read response call update safe-ts if there are no locks at this timestamp in leader.
+
+This feature would be enabled by default and no new configuration will be added. 
+
+### Protocol
+
+A new field `memory_lock` will be added to `ReadIndexContext` to send the lock status to replica.
+
+```
+pub struct ReadIndexContext {
+    pub id: Uuid,
+    pub request: Option<raft_cmdpb::ReadIndexRequest>,
+    pub locked: Option<LockInfo>,
+    pub memory_lock: Option<u64>,
+}
+```
+### Code changes
+
+Changes on follower
+```
+fn propose_raft_command {
+    RequestPolicy::ReadIndex => {
+        // try_local_stale_read
+        // if time stamp > safe-ts than send read index message to leader
+    }
+}
+
+fn apply_reads {
+    let read_index_ctx = ReadIndexContext::parse(state.request_ctx.as_slice()).unwrap();
+    if read_index_ctx.memory_lock == false {
+        // there are no pending locks on leader update safe-ts
+        // No need to take core lock as it is happening in raft thread
+        self.read_progress.update_safe_ts(apply_indx, ts)
```

Comment:
How the `ts` updating `safe_ts` on the follower is calculated? Is it the `read-ts` of the read request?

No memory locks indicates no pending write transctions, there could be still onoging transaction with prewrite locks in the lock-cf.

Reply:
Yes, safe TS can surpass resolve-TS with this change in follower. I think it should be OK to break the assumption that safe TS is always behind resolved TS in follower.

Reply:
Currently, `safe_ts` guarantees the **snapshot read** semantics, meaning no data with `commit_ts <= safe_ts` will be written anymore.

Although `read_ts` on the leader does not detect any memory locks, it only indicates that there are no ongoing async commit/1PC transactions. However, there may still exist prewrite locks in the lock-cf, and these prewrite locks might write data with `commit_ts < read_ts` during the commit process. Therefore, directly using `read_ts`, which only checks the leader memory lock, to update the follower's `safe_ts` could break the snapshot guarantees provided by `safe_ts` on the follower.

Reply:
If `read_ts` is to be used to update the follower peer's `safe_ts`, another approach is to update it not when the read index response is received, but after the follower read has been successfully processed. If the follower read completes correctly and returns results, it indicates that no conflicting prewrite locks were encountered on the follower. Combined with the known condition that no memory locks were read on the leader, this might ensure that no transactions with `commit_ts <= safe_ts` are written, thereby preserving the snapshot semantics.

Reply:
why safe_ts need to guarantees the snapshot read semantics.

Reply:
Can we say that 'safe-ts' is the timestamp that indicates up to which point the follower has committed data from the Raft state machine?

Reply:
The concepts of `safe_ts` and `follower-read` are indeed separate and serve different purposes.

- **Follower-read**: The read requests to be served as long as the `applied_index` on the follower is greater than the `commit_index` from the leader. It ensures that the data read is up-to-date with the leader’s committed state but doesn’t directly guarantee future write states.

- **`safe_ts` semantics**: The `safe_ts` guarantees that a read with `read_ts <= safe_ts` satisfies snapshot semantics. This means no transactions with `commit_ts <= safe_ts` will be written in the future.  

In the scenario described:  
- If there’s a **prewrite lock with `start_ts = 1`**, a `read_ts = 3` cannot be used to update the follower’s `safe_ts`, even if no memory lock is observed on the leader. The reason is that the prewrite lock might eventually commit a record with `commit_ts = 2`, which violates the snapshot semantics if the `safe_ts` had already been advanced to 3.

Therefore, advancing `safe_ts` requires additional safeguards to ensure no transactions with `commit_ts < read_ts` can still be committed after the `safe_ts` is updated. This is why relying solely on the absence of memory locks in the leader when using `read_ts` to update `safe_ts` on follower peer is insufficient.

Reply:
> Can i introduce apply-ts which will be updated by read index.

This introduces more concepts and changes. If the current goal is not to completely refactor the stale read design to achieve the x00ms target, there’s no need to make it this complex.  

If the **goal** is to reuse the information in the read index response to advance `safe_ts`, the mentioned approach can be used already:  
1. If the read index response does not contain **memory locks**, it indicates that the leader currently has no ongoing async commit transactions writing.  
2. If the follower read can process this request successfully at `read_ts` without returning a "key is locked" error, it means there are no conflicting prewrite locks at `read_ts`.  

If both conditions are met, the follower peer can consider using `read_ts` to update `safe_ts` directly.

Reply:
I agree that introducing new ts complicates the system. Ideally I want to update a ts (lets call it last-read-index ts) as soon as it get the read index response because we have all the information to serve the request with ts < last-read-index ts locally. It will help in two ways 
1. It allows more read index requests to be served locally
2. In case of retry, we might avoid retrying all the read index request if there is a missing response in between.

Having said that we can do incremental development  
1. We can experiment with updating safe-ts only after follower read is successfully read locally.
2. If above doesn't give better performance, we can introduce a new semantic last-read-index-ts. I don't think there will be require too many code changes.

Reply:
If the goal is solely to optimize follower reads by allowing more read index requests to be served locally, introducing `last-read-index` seems fine. It hepls keeping the optimization confined to follower reads avoids coupling with `safe_ts` or `resolved_ts` and introduce risks to safe_ts-related functionality.

Reply:
incorporated the review comments